File: backend/app/contracts/llm_profile_contract.py
"""LLM-Profile-Contract (Pydantic v2) — P5.1 Layer 0.

Repräsentiert benannte LLM-Konfigurationen (Provider + Modell + Key).
Jeder Agora-Run kann pro Schritt ein Profil referenzieren.
"""
from __future__ import annotations

from datetime import datetime
from typing import Optional

from pydantic import BaseModel, ConfigDict, Field

from .provider_types import ProviderType

_STRICT = ConfigDict(extra="forbid", populate_by_name=True)


class LlmProfile(BaseModel):
    model_config = _STRICT

    id: str = Field(..., description="UUID, server-generiert")
    name: str = Field(..., min_length=1, max_length=80)
    provider: ProviderType
    base_url: str = Field(..., min_length=1)
    model_name: str = Field(..., min_length=1)
    # None = nicht gesetzt (Update lässt Feld weg). "" = explizit geleert.
    api_key: Optional[str] = None
    is_default: bool = False
    created_at: datetime
    updated_at: datetime


class LlmProfileListResponse(BaseModel):
    model_config = _STRICT
    profiles: list[LlmProfile]


class LlmProfileCreateRequest(BaseModel):
    model_config = _STRICT

    name: str = Field(..., min_length=1, max_length=80)
    provider: ProviderType
    base_url: str = Field(..., min_length=1)
    model_name: str = Field(..., min_length=1)
    # None = nicht gesetzt (Update lässt Feld weg). "" = explizit geleert.
    api_key: Optional[str] = None
    is_default: bool = False

File: frontend/src/contracts/llmProfileContract.ts
/**
 * LLM-Profile-Contract — Zod-Spiegel zu backend/app/contracts/llm_profile_contract.py
 * P5.1 Layer 0. Spiegelt LlmProfile, LlmProfileListResponse, LlmProfileCreateRequest.
 */
import { z } from "zod";

export const LlmProviderSchema = z.enum([
  "ollama",
  "openai",
  "google",
  "gemini",
  "anthropic",
  "custom",
  "ollama_cloud",
  "openai_compatible",
  "minimax",
  "opencode_go",
  "github_copilot",
  "bedrock",
  "cloud",
  "codex_cli",
  "unknown",
]);
export type LlmProvider = z.infer<typeof LlmProviderSchema>;

export const LlmProfileSchema = z
  .object({
    id: z.string(),
    name: z.string().min(1).max(80),
    provider: LlmProviderSchema,
    base_url: z.string().min(1),
    model_name: z.string().min(1),
    // null = nicht gesetzt (Update lässt Feld weg). "" = explizit geleert.
    api_key: z.string().nullable().default(null),
    is_default: z.boolean().default(false),
    created_at: z.string(), // ISO-String (datetime serialisiert als str)
    updated_at: z.string(),
  })
  .strict();
export type LlmProfile = z.infer<typeof LlmProfileSchema>;

export const LlmProfileListResponseSchema = z
  .object({
    profiles: z.array(LlmProfileSchema),
  })
  .strict();
export type LlmProfileListResponse = z.infer<typeof LlmProfileListResponseSchema>;

export const LlmProfileCreateRequestSchema = z
  .object({
    name: z.string().min(1).max(80),
    provider: LlmProviderSchema,
    base_url: z.string().min(1),
    model_name: z.string().min(1),
    api_key: z.string().nullable().default(null),
    is_default: z.boolean().default(false),
  })
  .strict();
export type LlmProfileCreateRequest = z.infer<typeof LlmProfileCreateRequestSchema>;
